design/ratelimit-design.md

```diff
@@ -0,0 +1,154 @@
+
+Initially support a subset of the rate limiting features available by looking to implement rate limiting via a `key` which will be automatically determined by Contour for that route.
+Contour will support rate limiting via the `remote_address` presented to Envoy generically as well as allowing users to specify an IP address which will allow for blocking specific IPs (by defining a `requests_per_unit` of zero).
+
+## Detailed Design
+
+### Lyft RateLimit Implementation
+
+___Note: Much of the following documentation is from Lyft's implementation [github repo](https://github.com/lyft/ratelimit), but copied here for completeness of this design document.___
+
+The rate limit service is a Go/gRPC service designed to enable generic rate limit scenarios from different types of applications. Applications request a rate limit decision based on a domain and a set of descriptors. The service reads the configuration from disk via runtime, composes a cache key, and talks to the Redis cache. A decision is then returned to the caller.
+
+Contour will configure the domain it uses to be `contour` by default, but a new argument will be added which will allow users to customize this value.
+
+#### Descriptor list definition
+
+Each configuration contains a top level descriptor list and potentially multiple nested lists beneath that. The format is:
+
+```
+domain: <unique domain ID>
+descriptors:
+  - key: <rule key: required>
+    value: <rule value: optional>
+    rate_limit: (optional block)
+      unit: <see below: required>
+      requests_per_unit: <see below: required>
+    descriptors: (optional block)
+      - ... (nested repetition of above)
+```
+
+#### Rate limit definition
+```
+rate_limit:
+  unit: <second, minute, hour, day>
+  requests_per_unit: <uint>
+```
+
+The rate limit block specifies the actual rate limit that will be used when there is a match. Currently the service supports per second, minute, hour, and day limits.
+
+#### Example Configurations
+
+Limit all requests based upon a defined key:
+
+```
+---
+domain: contour
+descriptors:
+  - key: generic_key
+    value: apis  # This key will be configured in the annotation or IngressRoute
+    rate_limit:
+      unit: minute
+      requests_per_unit: 1
+```
+
+Limit requests based upon the remote_address:
+```
+---
+domain: contour
+descriptors:
+  - key: remote_address
+    rate_limit:
+      unit: second
+      requests_per_unit: 10
+```
+
+Block all requests from a specific IP:
+```
+---
+domain: contour
+descriptors:
+  - key: remote_address
+    value: 50.0.0.5
+    rate_limit:
+      unit: second
+      requests_per_unit: 0
+```
+
+### IngressRoute Design
+
+The `IngressRoute` spec will be modified to allow specific routes to enable rate limiting.
+A new struct will be added to allow users to define what type of rate limiting should be applied to the route.
+
+The `rateLimit` struct will have two possible values:
```

Comment:
This feels like an envoy implementation detail, is there a way we can infer this value from the rate limit parameters supplied by the user?

Reply:
Thought about this some more, I made this a map[string]string so that users could define whatever values they want. I think this is better and a good catch on the design.